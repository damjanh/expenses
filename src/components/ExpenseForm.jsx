import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpsenseForm extends React.Component {
    constructor(props) {
        super(props);

        const {
            expense,
        } = props;

        // eslint-disable-next-line react/state-in-constructor
        this.state = {
            description: expense.description ? expense.description : '',
            note: expense.note ? expense.note : '',
            amount: expense.amount ? (expense.amount / 100).toString() : '',
            createdAt: expense.createdAt ? moment(expense.createdAt) : moment(),
            calendarFocused: false,
            error: '',
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }

    onDateChange = (date) => {
        if (date) {
            this.setState(() => ({ createdAt: date }));
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }

    onSubmit = (e) => {
        e.preventDefault();
        const {
            description,
            amount,
            note,
            createdAt,
        } = this.state;
        if (!description || !amount) {
            this.setState(() => ({ error: 'Please provide a description and amount.' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description,
                note,
                amount: parseFloat(amount, 10) * 100,
                createdAt: createdAt.valueOf(),
            });
        }
    }

    render() {
        const {
            description,
            note,
            amount,
            createdAt,
            calendarFocused,
            error,
        } = this.state;
        return (
            <div>
                {error && <p>{error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        value={description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        id="sdp1"
                        onDateChange={this.onDateChange}
                        date={createdAt}
                        focused={calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note for your expense. (Optional)"
                        value={note}
                        onChange={this.onNoteChange}
                    />
                    <button type="submit">Add Expense</button>
                </form>
            </div>
        );
    }
}

ExpsenseForm.defaultProps = {
    expense: {
        description: '',
        createdAt: moment(),
        amount: '',
        note: '',
    },
};

ExpsenseForm.propTypes = {
    expense: PropTypes.shape({
        description: PropTypes.string,
        createdAt: PropTypes.string,
        amount: PropTypes.string,
        note: PropTypes.string,
    }),
    onSubmit: PropTypes.func.isRequired,
};
