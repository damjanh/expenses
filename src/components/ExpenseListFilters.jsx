import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {
    setTextFilter,
    sortByAmount,
    sortByDate,
    setStartDate,
    setEndDate,
} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };

    onDatesChange = ({ startDate, endDate }) => {
        const { setStartDateProp, setEndDateProp } = this.props;
        setStartDateProp(startDate);
        setEndDateProp(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    onTextChange = (e) => {
        const { setTextFilterProp } = this.props;
        setTextFilterProp(e.target.value);
    }

    onSortChange = (e) => {
        const { sortByDateProp, sortByAmountProp } = this.props;
        if (e.target.value === 'date') {
            sortByDateProp();
        } else if (e.target.value === 'amount') {
            sortByAmountProp();
        }
    }

    render() {
        const {
            filters: {
                text,
                sortBy,
                startDate,
                endDate,
            },
        } = this.props;
        const { calendarFocused } = this.state;
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            className="text-input"
                            placeholder="Search expenses"
                            type="text"
                            value={text}
                            onChange={this.onTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={sortBy}
                            onChange={this.onSortChange}
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={startDate}
                            endDate={endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            startDateId="start"
                            endDateId="end"
                            showClearDates
                        />
                    </div>
                </div>
            </div>
        );
    }
}

ExpenseListFilters.propTypes = {
    setTextFilterProp: PropTypes.func.isRequired,
    sortByDateProp: PropTypes.func.isRequired,
    sortByAmountProp: PropTypes.func.isRequired,
    setStartDateProp: PropTypes.func.isRequired,
    setEndDateProp: PropTypes.func.isRequired,
    filters: PropTypes.shape({
        text: PropTypes.string.isRequired,
        sortBy: PropTypes.string.isRequired,
        startDate: PropTypes.object,
        endDate: PropTypes.object,
    }).isRequired,
};

const mapDispatchToProps = (dispatch) => (
    {
        setTextFilterProp: (text) => dispatch(setTextFilter(text)),
        sortByDateProp: () => dispatch(sortByDate()),
        sortByAmountProp: () => dispatch(sortByAmount()),
        setStartDateProp: (startDate) => dispatch(setStartDate(startDate)),
        setEndDateProp: (endDate) => dispatch(setEndDate(endDate)),
    }
);

const mapStateToProps = (state) => (
    {
        filters: state.filters,
    }
);

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
