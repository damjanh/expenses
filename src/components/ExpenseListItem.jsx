import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = (props) => {
    const {
        id,
        description,
        amount,
        createdAt,
        dispatch,
    } = props;
    return (
        <div>
            <h3>{description}</h3>
            <p>
                {amount}
                -
                {createdAt}
            </p>
            <button
                type="button"
                onClick={() => {
                    dispatch(removeExpense({ id }));
                }}
            >
                Remove
            </button>
        </div>
    );
};

ExpenseListItem.propTypes = {
    dispatch: PropTypes.func.isRequired,
    id: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    amount: PropTypes.number.isRequired,
    createdAt: PropTypes.number.isRequired,
};

export default connect()(ExpenseListItem);
