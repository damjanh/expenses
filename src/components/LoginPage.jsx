import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import startLogin from '../actions/auth';

export const LoginPage = (props) => {
    const { startLoginProp } = props;
    return (
        <div>
            <button type="button" onClick={startLoginProp}>
                Login
            </button>
        </div>
    );
};

LoginPage.propTypes = {
    startLoginProp: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
    startLoginProp: () => dispatch(startLogin()),
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
