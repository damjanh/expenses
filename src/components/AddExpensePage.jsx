import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        const { history, startAddExpenseProp } = this.props;
        startAddExpenseProp(expense);
        history.push('/');
    };

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}

AddExpensePage.propTypes = {
    startAddExpenseProp: PropTypes.func.isRequired,
    history: PropTypes.shape({ push: PropTypes.func.isRequired }).isRequired,
};

const mapDispatchToProps = (dispatch) => ({
    startAddExpenseProp: (expense) => dispatch(startAddExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
