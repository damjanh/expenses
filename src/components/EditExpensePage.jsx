import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        const { startEditExpenseProp, history, expense: { id } } = this.props;
        startEditExpenseProp(id, expense);
        history.push('/');
    }

    onRemove = () => {
        const { startRemoveExpenseProp, history, expense: { id } } = this.props;
        startRemoveExpenseProp(id);
        history.push('/');
    }

    render() {
        const {
            expense,
        } = this.props;
        return (
            <div>
                <ExpenseForm
                    expense={expense}
                    onSubmit={this.onSubmit}
                />
                <button
                    type="submit"
                    onClick={this.onRemove}
                >
                    Remove
                </button>
            </div>
        );
    }
}

EditExpensePage.defaultProps = {
    expense: {
        note: '',
    },
};

EditExpensePage.propTypes = {
    startEditExpenseProp: PropTypes.func.isRequired,
    startRemoveExpenseProp: PropTypes.func.isRequired,
    history: PropTypes.shape({ push: PropTypes.func.isRequired }).isRequired,
    expense: PropTypes.shape({
        id: PropTypes.string.isRequired,
        createdAt: PropTypes.number.isRequired,
        description: PropTypes.string.isRequired,
        amount: PropTypes.number.isRequired,
        note: PropTypes.string,
    }),
};

const mapStateToProps = (state, props) => {
    const expense = state.expenses.find((exp) => exp.id === props.match.params.id);
    return {
        expense,
    };
};

const mapDispatchToProps = (dispatch) => ({
    startEditExpenseProp: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpenseProp: (id) => dispatch(startRemoveExpense(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
