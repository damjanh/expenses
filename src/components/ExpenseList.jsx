/* eslint-disable react/prop-types */
import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';

const ExpenseList = (props) => {
    const { expenses } = props;
    return (
        <div>
            <h1>Expense List</h1>
            {expenses.map((expense) => (
                <ExpenseListItem
                    key={expense.id}
                    description={expense.description}
                    amount={expense.amount}
                    createdAt={expense.createdAt}
                />
            ))}
        </div>
    );
};

const mapStateToProps = (state) => ({
    expenses: state.expenses,
});

export default connect(mapStateToProps)(ExpenseList);
