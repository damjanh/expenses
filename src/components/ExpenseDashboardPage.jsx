/* eslint-disable import/no-named-as-default */
import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';
/* eslint-enable import/no-named-as-default */

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;
