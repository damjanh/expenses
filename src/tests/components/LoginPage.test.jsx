import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';

test('Should render LoginPage correctly', () => {
    const wrapper = shallow(<LoginPage startLoginProp={() => {}} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should call startLogin when login button is clicked', () => {
    const startLogin = jest.fn();
    const wrapper = shallow(<LoginPage startLoginProp={startLogin} />);
    wrapper.find('button').simulate('click');
    expect(startLogin).toHaveBeenCalled();
});
