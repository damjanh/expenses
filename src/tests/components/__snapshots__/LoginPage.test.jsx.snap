// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render LoginPage correctly 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      Expenses
    </h1>
    <p>
      Control your expenses!
    </p>
    <button
      className="button"
      onClick={[Function]}
      type="button"
    >
      Login with Google
    </button>
  </div>
</div>
`;
